Logging framework

tchamberlin opened the pull request:
This PR introduces a logging framework to Dysh, via the `logging` module from the standard library. It will definitely need some tweaks before it goes live, but hopefully it is pretty close to done.

It also includes a _limited_ conversion of the existing `if debug: ...` style logging statements to using `logging.debug`.

The core logging is set up in `dysh.log`. Its responsibilities are:

1. Configure logging. This is done via `logging.config.dictConfig`
2. Provide access to "the" dysh logger, `"dysh"`. This avoids having to do `import logging; logger = logging.getLogger("dysh")` in every module, which is error prone and noisy. Instead, modules just need to `from dysh.log import logger`

There are three different log handlers:

1. `stderr`: Logs DEBUG+ to `stderr` using a "simple" formatter. Configured by default
2. `dysh_global_log_file`: Logs INFO+ to a file, which defaults to `$DYSH_HOME/dysh.log`. `DYSH_HOME` defaults to `.` currently, because figuring out this sort of thing seems out of scope for this PR. This file is _appended_ to. Configured by default
3. `dysh_instance_log_file`: An _optional_ file logger which creates a new file for each execution; DEBUG+. It is used e.g. via `$dysh --log PATH`

I don't know that we actually want to use all three of these, but I think at least two of them. This is open for discussion

The actual log level can be configured via `dysh.log.config_logging`

## Usage



### CLI Usage

Here's an example that demonstrates the new (converted) logging from the FITS loaders:

```plain
$ dysh -v3 --no-banner /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas --log "./$(date +"%Y-%m-%d_%H-%M-%S").log" 
DEBUG Logging has been set to verbosity 3 / level DEBUG
INFO Log file for this instance of dysh: 2024-08-08_10-45-17.log
DEBUG Treating given path /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas as a directory
DEBUG Selecting /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.A.fits to load
DEBUG Loading HDU 1 from /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.A.fits
DEBUG Selecting /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.B.fits to load
DEBUG Loading HDU 1 from /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.B.fits
DEBUG Selecting /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.C.fits to load
DEBUG Loading HDU 1 from /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.C.fits
DEBUG Selecting /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.D.fits to load
DEBUG Loading HDU 1 from /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.D.fits
Loaded 4 FITS files
```

### Notebook Usage

![image](https://github.com/user-attachments/assets/e872951f-5b74-48ab-9ef0-43c992af85fa)

## Instance Log File Example

```plain
2024-08-08T12:02:42-0400 - log.config_logging - DEBUG - Logging has been set to verbosity 3 / level DEBUG
2024-08-08T12:02:42-0400 - gbtfitsload.__init__ - DEBUG - Treating given path /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas as a directory
2024-08-08T12:02:42-0400 - gbtfitsload.__init__ - DEBUG - Selecting /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.A.fits to load
2024-08-08T12:02:42-0400 - sdfitsload.load - DEBUG - Loading HDU 1 from /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.A.fits
2024-08-08T12:02:42-0400 - gbtfitsload.__init__ - DEBUG - Selecting /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.B.fits to load
2024-08-08T12:02:42-0400 - sdfitsload.load - DEBUG - Loading HDU 1 from /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.B.fits
2024-08-08T12:02:42-0400 - gbtfitsload.__init__ - DEBUG - Selecting /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.C.fits to load
2024-08-08T12:02:42-0400 - sdfitsload.load - DEBUG - Loading HDU 1 from /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.C.fits
2024-08-08T12:02:43-0400 - gbtfitsload.__init__ - DEBUG - Selecting /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.D.fits to load
2024-08-08T12:02:43-0400 - sdfitsload.load - DEBUG - Loading HDU 1 from /home/dysh/public_html/example_data/subbeamnod-Ka/data/TRCO_230413_Ka.raw.vegas/TRCO_230413_Ka.raw.vegas.D.fits
```

mpound commented:
Overall this is excellent.   A few comments:

1. Is there a way to turn off logging to stderr when instance logging is going to a file? Currently any message goes to screen and log file.

```
% dysh -v3 /home/mpound/src/dysh/testdata/AGBT18B_354_03/AGBT18B_354_03.raw.vegas/ --log test.log
In [1]: from dysh.log import config_logging,logger
In [2]: logger.critical("this is critical!")
CRITICAL this is critical!
In [3]: !cat test.log
2024-08-08T10:58:02-0700 - CRITICAL - <ipython-input-2-c628d8f3efab> - this is critical!
```

2. A warning is given when switching instance log files:
```
In [4]: config_logging(verbosity=2,path="./test2.log")
INFO Log file for this instance of dysh: ./test2.log
/home/mpound/Anaconda3-2023.03/envs/3.9-env/lib/python3.9/logging/config.py:783: ResourceWarning: unclosed file <_io.TextIOWrapper name='/home/mpound/src/dysh/dysh.log' mode='a' encoding='ISO-8859-1'>
  for h in logger.handlers[:]:
/home/mpound/Anaconda3-2023.03/envs/3.9-env/lib/python3.9/logging/config.py:795: ResourceWarning: unclosed file <_io.TextIOWrapper name='/home/mpound/src/dysh/test.log' mode='w' encoding='ISO-8859-1'>
  self.common_logger_config(logger, config, incremental)
```

This warning only happens if ipython is started with `dysh`;  it does not happen if started with just `ipython` 
```
% ipython
In [1]: from dysh.log import logger, config_logging
In [2]: config_logging(path="/tmp/foobar.log",verbosity=3)
DEBUG Logging has been set to verbosity 3 / level DEBUG
In [3]: logger.debug("testing 456")
DEBUG testing 456
In [4]: config_logging(path="/tmp/foobar2.log",verbosity=3)
INFO Log file for this instance of dysh: /tmp/foobar2.log
DEBUG Logging has been set to verbosity 3 / level DEBUG
In [5]: logger.debug("testing 890")
DEBUG testing 890
```

tchamberlin commented:
Well first of all let me clarify the goals here. We want to support two different use cases:

1. Dysh is used as an application, e.g. via `$ dysh`, or the GUI. In these cases, we want to automatically configure logging for the user. This means that we need to set up handlers, formatters, loggers, etc., and then provided some limited configuration over these via CLI flag, config file, etc. Right now that is just `--verbosity` and `--log` args to `$ dysh`
2. Dysh is used as a library, e.g.  via `$ dysh-lab`, or as part of a larger data reduction script that uses `dysh` as a dependency. In these cases we do _not_ want to automatically configure any logging. But we still want to be emitting log statements, so that if users of dysh-the-library opt into logging they will be able to get useful information from it

For Application usage, users should not be calling `config_logging`, as it has already been called for them. So I would expect some strange behavior in these cases. I'm not sure if a `FileHandler` defines behavior in the scenario you describe; my suspicion is that the correct way to do what you are doing is to remove the existing handler and add a new one with a different path

To fix this, perhaps I could just add a check to `config_logging` that raises an error if logging has already been configured?


Regarding the configuration of separate levels for file and console output: yes, I think this is a good idea. I suggest adding a `-q/--quiet` argument, which is a pretty standard way of doing this I think. Typically though error-level output will _always_ be shown in the console, so we'd want to make sure we replicate that.

If we take `rsync` as a specific example...

```sh

$ # No output of any kind
$ rsync --dry-run /tmp/a /tmp/b

$ # Verbose output to console
$ rsync -v --dry-run /tmp/a /tmp/b
a
<snip>

$ # VERY verbose output to console
$ rsync -vvv --dry-run /tmp/a /tmp/b
<snip>

$ # Verbose output to console and log file
$ rsync -v --dry-run /tmp/a /tmp/b --log-file /tmp/rsync.log

$ # Verbose output to log file, but only error-level output to console
$ rsync -q -vvv --dry-run /tmp/nope /tmp/b --log-file /tmp/rsync.log
rsync: link_stat "/tmp/nope" failed: No such file or directory (2)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1189) [sender=3.1.3]
```

I think this is the same pattern you are suggesting?

mpound commented:
Adding  a check to `config_logging` that raises an error if logging has already been configured is a good solution.

The non-application, the user will still need to use config_logging, correct?  They will also need a way to shush the stderr handler. Maybe adding a quiet boolean to config_logging?

I like your analog with rsync; that would work great with dysh-the-application and dysh-lab

tchamberlin commented:
I've added the `--quiet` flag, renamed `config_logging` to `init_logging` to better reflect its purpose, and added the warning about re-running `init_logging`. Demo:

With `--quiet`:

```plain
$ dysh --no-banner -v3 -q

In [1]: from dysh.log import logger, init_logging

In [2]: logger.info("This will appear in dysh.log, but not in stderr")

In [3]: cat dysh.log
2024-08-15T10:54:36-0400 - INFO - log - test
2024-08-15T10:55:40-0400 - WARNING - log - dysh logging has already been initialized! Continuing, but this behavior is not well defined
2024-08-15T10:56:51-0400 - WARNING - log - dysh logging has already been initialized! Continuing, but this behavior is not well defined, and you will likely end up with duplicate log handlers
2024-08-15T10:57:38-0400 - INFO - <ipython-input-2-505319171c0b> - This will appear in dysh.log, but not in stderr

In [4]: init_logging(3)
WARNING dysh logging has already been initialized! Continuing, but this behavior is not well defined, and you will likely end up with duplicate log handlers
```

Without `--quiet`:

```plain
$ dysh --no-banner -v3 
DEBUG Logging has been set to verbosity 3 / level DEBUG

In [1]: from dysh.log import logger, init_logging

In [2]: logger.info("This will appear in BOTH stderr and dysh.log")
INFO This will appear in BOTH stderr and dysh.log
```

mpound commented on src/dysh/shell/shell.py:
```diff
@@ -61,8 +65,8 @@ def init_shell(*ipython_args, colors=DEFAULT_COLORS, profile: Union[str, Path] =
     user_ns = {"pd": pd, "np": np, "GBTFITSLoad": GBTFITSLoad, "Table": Table, "fits": fits}
-    if sdfits_files:
-        user_ns["sdfits_files"] = sdfits_files
+    # if sdfits_files:
```
dysh xx.fits  loads xx.fits into a GBTFITSLoad to which the user has no access because lines 68-69 are commented out.   Is there a reason you did this?

mpound commented on src/dysh/shell/shell.py:
dysh xx.fits loads xx.fits into a GBTFITSLoad to which the user has no access because lines 68-69 are commented out. Is there a reason you did this?

tchamberlin replied:
This is a (partial) bug fix and doesn't really belong in this PR. Don't remember the exact error
